### ISSUE TITLE
cmd/go: download go1.22 for darwin/arm64: toolchain not available

### ISSUE THREAD

**Issue by EwenQuim** [golang/go#65568]
### Go version

go1.22

### Output of `go env` in your module/workspace:

```shell
go: downloading go1.22 (darwin/arm64)
go: download go1.22 for darwin/arm64: toolchain not available
```

### What did you do?

```
rm -rf /usr/local/go/
rm -rf ~/go
```

Then used the installer.

<img width="735" alt="image" src="https://github.com/golang/go/assets/46993939/3be2c953-e8f1-4f4e-93cd-2c41bb9e6eda">

Then ran the command `go version`.

```
go: downloading go1.22 (darwin/arm64)
go: download go1.22 for darwin/arm64: toolchain not available
```

`go env` also returns this result.

<img width="465" alt="image" src="https://github.com/golang/go/assets/46993939/92e62167-a59e-4a28-81dd-e18dc97cbd18">

### What did you see happen?

`go: download go1.22 for darwin/arm64: toolchain not available`, not much more

### What did you expect to see?

go1.22

**Comment by dmitshur** [comment]
Thanks for reporting. What's the output of `which go` and `GOTOOLCHAIN=local go env`?

**Comment by adamdecaf** [comment]
I've installed Go with homebrew and am on an Intel mac. 

```
go: downloading go1.22 (darwin/amd64)
go: download go1.22 for darwin/amd64: toolchain not available
```

```
$ which go
/usr/local/bin/go

$ GOTOOLCHAIN=local go env
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/Users/adam/Library/Caches/go-build'
GOENV='/Users/adam/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/adam/code/pkg/mod'
GONOPROXY='github.com/moov-io/*,github.com/moovfinancial/*'
GONOSUMDB='github.com/moov-io/*,github.com/moovfinancial/*'
GOOS='darwin'
GOPATH='/Users/adam/code/'
GOPRIVATE='github.com/moov-io/*,github.com/moovfinancial/*'
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/Cellar/go/1.21.6/libexec'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/local/Cellar/go/1.21.6/libexec/pkg/tool/darwin_amd64'
GOVCS=''
GOVERSION='go1.21.6'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='cc'
CXX='c++'
CGO_ENABLED='1'
GOMOD='/Users/adam/code/src/github.com/moov-io/irs/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/5c/9jlzxrmd1_7_fw3d05nb62dc0000gn/T/go-build2824011357=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

**Comment by dmitshur** [comment]
@adamdecaf Since your Go toolchain was installed using a different tool, it's hard to know if it's the same problem or a different homebrew-specific problem. Let's wait to hear back from @EwenQuim or someone who's able to reproduce this the problem that affects the Go installer from go.dev/dl/.

**Comment by EwenQuim** [comment]
`which go` : `/usr/local/go/bin/go`

<details>
<summary>GOTOOLCHAIN=local go env</summary>
<pre>
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/ewenq/Library/Caches/go-build'
GOENV='/Users/ewenq/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/ewenq/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/ewenq/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/local/go/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.22.0'
GCCGO='gccgo'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/5g/q9_gydg96zzd59zgh0hscvqh0000gn/T/go-build2955048168=/tmp/go-build -gno-record-gcc-switches -fno-common'
</pre>
</details>

I don't see anything wrong here, but it might be only my local setup

**Comment by seankhliao** [comment]
is this in a module (if yes, what's the `go` directive in there)?
what's the `GOTOOLCHAIN` env set to?
what's in `/Users/ewenq/Library/Application Support/go/env`?

**Comment by dmitshur** [comment]
@EwenQuim Thanks.

Can you post more complete steps for how to reproduce it, in an empty directory?

For example, does it reproduce if you do the following sequence?

```
cd $(mktemp -d)
go version
```

Is the `GOTOOLCHAIN` variable set in your environment?

I can only reproduce this if I set `GOTOOLCHAIN` to explicitly to "go1.22" (which is not a Go toolchain that exists):

```
$ cd $(mktemp -d)
$ GOTOOLCHAIN=go1.22 go version
go: downloading go1.22 (darwin/arm64)
^C
```

Otherwise it doesn't reproduce:

```
$ cd $(mktemp -d)
$ echo $GOTOOLCHAIN

$ go version
go version go1.22.0 darwin/arm64
```

**Comment by EwenQuim** [comment]
@seankhliao 

> is this in a module (if yes, what's the go directive in there)?

No

> what's the GOTOOLCHAIN env set to?

Nothing
<img width="157" alt="image" src="https://github.com/golang/go/assets/46993939/e28ebf21-aab0-49d7-b460-7cc1b2e06862">


> what's in /Users/ewenq/Library/Application Support/go/env?
<img width="385" alt="image" src="https://github.com/golang/go/assets/46993939/5083e951-8c36-4542-9428-8fd3cb54aac7">

**Comment by EwenQuim** [comment]
OK the problem is solved after I deleted the `/Users/ewenq/Library/Application Support/go/env` file, thank you very much !

Sorry for taking some of your time, I will explain the cause of the bug below if you're interested, but no need to search more.

**Comment by dmitshur** [comment]
Thanks. It's good to understand that the GOTOOLCHAIN=go1.22 line in the go/env file was causing this, and that it's not a problem with the Go installer.

A problem like that is fixable, as you've done, by unsetting the unwanted value with `GOTOOLCHAIN=local go env -u GOTOOLCHAIN` or by manually editing (or deleting) that file.

**Comment by EwenQuim** [comment]
The invalid version of `GOTOOLCHAIN` that was causing me the same issue again and again even after deleting and reinstalling the `/usr/local/go` directory.

Good learning for me: the `/usr/local/go` directory isn't the only place Go config lives on my computer.

Maybe that updating the [go.dev](https://go.dev/doc/manage-install#uninstalling) page with more information can be a good idea to prevent future issues like this one ?

<img width="1268" alt="image" src="https://github.com/golang/go/assets/46993939/a81e22c6-b619-4f8c-87d2-4b3779be9836">

**Comment by kfelter** [comment]
Fixed it by editing the `go.mod` file
~go 1.22~
**go 1.22.0**

Then Quit and Reopened VS Code

**Comment by ldemailly** [comment]
Thanks for that last comment

I think it's very counter intuitive for `go version` to not output the actual binary version but instead start doing something else that depends on files in the current directory. I can't think of any other unix binary doing that (there is also no debug or verbose I can't see that explains that behavior, like "using go version from go.mod" output somewhere, after printing the actual current binary version

seems you have to do crazy stuff like
```
DIR=$(mktemp -d); cd $DIR; go version
go version go1.21.8 darwin/arm64
```
to get the actual go version instead of
```
$ go version
go: downloading go1.22 (darwin/arm64)
go: download go1.22 for darwin/arm64: toolchain not available
```

also probably should say "go.mod has go 1.22 which isn't semantic versioning, please correct to 1.22.0"
or fix the behavior so 1.22 means... like it used to... "latest release of 1.22" instead of failing?

**Comment by dmitshur** [comment]
> you have to do [...] `DIR=$(mktemp -d); cd $DIR; go version`

If you'd like to determine the bundled Go toolchain's version, it's possible to do it with `GOTOOLCHAIN=local go version`. This is documented in the last paragraph of https://go.dev/doc/toolchain#select. The rest of that document provides useful context for the current behavior.

> also probably should say "go.mod has go 1.22 which isn't semantic versioning, please correct to 1.22.0"
or fix the behavior so 1.22 means... like it used to... "latest release of 1.22" instead of failing?

@ldemailly You're right that this can be improved. This is now tracked in issue #66175. Please feel free to follow it for updates.

Since this issue is closed and there's another open one to track this problem, I'll lock this to avoid further replies pinging many people. If someone would like to report a different problem, please file a new issue at [go.dev/issue/new](https://go.dev/issue/new). Thanks.
